File: 2021/day4/scala/solution.scala
import scala.io.Source

type Board = List[List[Int]]
type Row = List[Int]
type Col = List[Int]
type Draws = List[Int]

def part1(path: String) = {
  val (draws, boards) = parseInput(path)

  var seenDraws = draws.take(4)
  var winningBoard : Option[Board] = None
  while winningBoard == None do 
    seenDraws = draws.take(seenDraws.length + 1)
    winningBoard = boards.find(isWinningBoard(_, seenDraws))

  val unmarkedNumbers = winningBoard.get
    .flatMap(row => row.filterNot(v => seenDraws.contains(v)))
  
  println(seenDraws.last * unmarkedNumbers.sum)
}


def part2(path: String) = {
  var (draws, boards) = parseInput(path)

  // Go until all but one board have won
  var seenDraws = draws.take(4)
  while boards.length > 1 do 
    seenDraws = draws.take(seenDraws.length + 1)
    boards = boards.filterNot(isWinningBoard(_, seenDraws))
  
  // Draw until the last board has won
  while (!isWinningBoard(boards.last, seenDraws)) do
    seenDraws = draws.take(seenDraws.length + 1)

  var unmarkedNumbers = boards.last
    .flatMap(row => row.filterNot(v => seenDraws.contains(v)))
  
  println(seenDraws.last * unmarkedNumbers.sum)
}

def isWinningBoard(b: Board, d: Draws) =
  rows(b).exists(isWinningRow(_, d)) || cols(b).exists(isWinningCol(_, d))
def isWinningRow(r: Row, d: Draws) = r.forall(d.contains(_))
def isWinningCol(c: Col, d: Draws) = isWinningRow(c, d)
def rows(b: Board) : List[Row] = b
def cols(b: Board) : List[Col] = b.transpose


def parseInput(path: String) : (Draws, List[Board]) = {
  val lines = Source.fromFile(path)
  .getLines

  val draws = lines.take(1).flatMap(_.split(",")).map(_.toInt).toList

  val boards = lines.drop(1).grouped(6).map(board => 
    board.take(5) // Skip empty 6th line if there
          .map(row => row.split("\\s+")
                .filter(_.nonEmpty)
                .map(_.toInt)
                .toList)
          .toList
  ).toList

  (draws, boards)
}


@main
def main() = {
  println("Part 1: ")
  print("Example: ")
  part1("../example.txt")
  print("Input: ")
  part1("../input.txt")
  println()

  println("Part 2: ")
  print("Example: ")
  part2("../example.txt")
  print("Input: ")
  part2("../input.txt")
  println()
}

File: 2021/day4/ruby/solution.rb
input = File.readlines(__dir__ + '/../input.txt').map(&:strip)
example = File.readlines(__dir__ + '/../example.txt').map(&:strip)

class Cell
  attr_accessor :value, :marked

  def initialize(value, marked) = (@value, @marked = value, marked)
  def mark! = @marked = true
  def to_s = (marked ? "[#{value}]" : " #{value} ").ljust(4)
  def inspect = "Cell<value=#{value}, marked=#{marked}>"
end

class Board
  attr_accessor :rows, :bingo, :marked_values

  def initialize(rows) = (@rows, @bingo, @marked_values = rows, false, [])
  def columns = @cols ||= rows.transpose
  def finished? = (rows + columns).any? { _1.all?(&:marked) }
  def cells = rows.flatten
  def mark!(value)
    marked_cells = cells.select { _1.value == value }.map(&:mark!)
    @marked_values << value if marked_cells.any?
    @bingo = true if finished?
  end
  def bingo? = @bingo
  def sum
    cells.reject(&:marked).map(&:value).sum * @marked_values.last
  end

  def to_s
    rows.map { _1.map(&:to_s).join(' ') }.join("\n")
  end
end

def parse_board(string_rows)
  Board.new string_rows.map { _1.split(' ').map { |value| Cell.new(value.to_i, false) } }
end

def parse_boards_from_input(input)
  input
    .drop(2)
    .each_with_object([[]]) { |row, boards| row.empty? ? boards << [] : boards.last << row }
     .map { parse_board(_1) }
end

def part1(input)
  draws = input.first.split(',').map(&:to_i)
  boards = parse_boards_from_input(input)

  while not (boards.any?(&:bingo?)) and not draws.empty?
    draw = draws.shift
    boards.each do |board|
      board.mark!(draw)
    end
  end

  boards.find(&:bingo?).sum
end

def part2(input)
  draws = input.first.split(',').map(&:to_i)
  boards = parse_boards_from_input(input)

  while not boards.all?(&:bingo?) and not draws.empty?
    draw = draws.shift
    boards.each do |board|
      board.mark!(draw)
    end
    return boards.first.sum if boards.size == 1 and boards.first.bingo?
    boards.filter! { not _1.bingo? }
  end
end

puts "Part 1 (Example): #{part1(example)}"
puts "Part 1: #{part1(input)}"
puts "Part 2 (Example): #{part2(example)}"
puts "Part 2: #{part2(input)}"
